--- a/src/lib/log.c
+++ b/src/lib/log.c
@@ -286,7 +286,7 @@ char const *fr_syserror(int num)
 	/*
 	 *	XSI-Compliant version
 	 */
-#if !defined(HAVE_FEATURES_H) || !defined(__GLIBC__) || ((_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 500) && ! _GNU_SOURCE)
+#if !defined(HAVE_FEATURES_H) || !defined(__GLIBC__) || ((_POSIX_C_SOURCE >= 200112L || defined(_XOPEN_SOURCE)) && ! defined(_GNU_SOURCE))
 	ret = strerror_r(num, p, end - p);
 	if (ret != 0) {
# ifndef NDEBUG
